torch/nn/parallel/replicate.py

```diff
@@ -1,42 +1,73 @@
-            continue
-        seen_params.add(param)
-        param_copies = Broadcast(device_ids)(param)
-        for param_copy, remap in zip(param_copies, param_remap):
-            remap[param] = param_copy
-    for m in module.modules():
-        for buffer in m._buffers.values():
-            copies = comm.broadcast(buffer, device_ids)
-            for buf_copy, remap in zip(copies, param_remap):
-                remap[buffer] = buf_copy
-    return [_replicate_module(module, device_id, remap)
-            for device_id, remap in zip(device_ids, param_remap)]
+
+    devices = tuple(devices)
+    num_replicas = len(devices)
+
+    params = list(network.parameters())
+    param_indices = {param: idx for idx, param in enumerate(params)}
+    param_copies = Broadcast(devices)(*params)
+    if len(params) > 0:
+        param_copies = [param_copies[i:i+len(params)]
+                        for i in range(0, len(param_copies), len(params))]
+
+    buffers = _buffers(network)
+    buffer_indices = {buf: idx for idx, buf in enumerate(buffers)}
+    buffer_copies = comm.broadcast_coalesced(buffers, devices)
+
+    modules = list(network.modules())
+    module_copies = [[] for device in devices]
+    module_indices = {}
+
+    for i, module in enumerate(modules):
+        module_indices[module] = i
+        for j in range(num_replicas):
+            replica = module.__new__(type(module))
+            replica.__dict__ = module.__dict__.copy()
+            replica._parameters = replica._parameters.copy()
+            replica._buffers = replica._buffers.copy()
+            replica._modules = replica._modules.copy()
+            module_copies[j].append(replica)
+
+    for i, module in enumerate(modules):
+        for key, child in module._modules.items():
+            module_idx = module_indices[child]
+            for j in range(num_replicas):
+                replica = module_copies[j][i]
+                replica._modules[key] = module_copies[j][module_idx]
+        for key, param in module._parameters.items():
+            if param is None:
+                for j in range(num_replicas):
+                    replica = module_copies[j][i]
+                    replica._parameters[key] = None
+            else:
+                param_idx = param_indices[param]
+                for j in range(num_replicas):
+                    replica = module_copies[j][i]
+                    replica._parameters[key] = param_copies[j][param_idx]
+        for key, buf in module._buffers.items():
+            if buf is None:
+                for j in range(num_replicas):
+                    replica = module_copies[j][i]
+                    replica._buffers[key] = None
+            else:
+                buffer_idx = buffer_indices[buf]
+                for j in range(num_replicas):
+                    replica = module_copies[j][i]
+                    replica._buffers[key] = buffer_copies[j][buffer_idx]
+
+    return [module_copies[j][0] for j in range(num_replicas)]
```

Comment:
The first module is the outermost module (i.e. `list(network.modules())[0] == network`), which is the same module that the user passed as an argument to replicate.

This is the same behavior that replicate had before:
```
  replicate(network, devices) -> replicas
```
where replicas is a list of copies of the module `network` (i.e.`len(replicas) == len(devices)`)